src/Trinity.C/src/Network/Server/posix/TrinitySocketServer.cpp

```diff
@@ -142,55 +157,67 @@ namespace Trinity
             }
             if (addrinfop == NULL)
             {
+                Diagnostics::WriteLine(Diagnostics::LogLevel::Error, "Cannot bind the socket to an IP endpoint {0}.", port);
+                ShutdownSocketServer();
                 return -1;
             }
-            printf("accept_sock: %d\n", accept_sock);
             //make_nonblocking(accept_sock);
             freeaddrinfo(addrinfos);
 
             if (-1 == listen(accept_sock, SOMAXCONN))
             {
-                printf("listen failed\n");
+                Diagnostics::WriteLine(Diagnostics::LogLevel::Error, "Cannot turn the socket into listening state.");
+                ShutdownSocketServer();
                 return -1;
             }
 
-            fprintf(stderr, "listen succeed\n");
-
-            if (-1 == InitializeEventMonitor())
+            if (0 != InitializeEventMonitor())
             {
-                close(accept_sock);
+                Diagnostics::WriteLine(Diagnostics::LogLevel::Error, "Cannot initialize socket server event monitor.");
+                ShutdownSocketServer();
                 return -1;
             }
 
-            socket_accept_thread = std::thread(SocketAcceptThreadProc);
+            Diagnostics::WriteLine(Diagnostics::LogLevel::Info, "Listening endpoint :{0}", port);
+
+            Diagnostics::WriteLine(Diagnostics::LogLevel::Info, "Waiting for client connection ...");
+
+            socket_accept_thread = new std::thread(SocketAcceptThreadProc);
 
             return accept_sock;
         }
 
         int ShutdownSocketServer()
         {
+            /* signal the accept thread that we are shutting down */
             shutdown(accept_sock, SHUT_RD);
-            socket_accept_thread.join();
+            if (socket_accept_thread != nullptr)
+            {
+                socket_accept_thread->join();
+                delete socket_accept_thread;
+                socket_accept_thread = nullptr;
+            }
             close(accept_sock);
             UninitializeEventMonitor();
-            // TODO close existing connections
-            return 0;
-        }
 
-        int AcceptConnection(int sock_fd)
-        {
-            fprintf(stderr, "waiting for connection ...\n");
-            sockaddr addr;
-            socklen_t addrlen = sizeof(sockaddr);
-            return accept4(sock_fd, &addr, &addrlen, SOCK_NONBLOCK);
+            while (g_threadpool_size > 0) { usleep(100000); }
+            /* Now, all WorkerThreadProc exit. Proceed to close all client sockets. */
+            psco_spinlock.lock();
+            std::vector<PerSocketContextObject*> psco_vec = std::vector<PerSocketContextObject*>(psco_map.size());
```

Comment:
To make sure everything goes atomically, and to avoid writing CloseClientConnection again.